src/submatrix_dissection.F

```diff
@@ -0,0 +1,895 @@
+               DO WHILE (l .LE. l_limit_right)
+                  IF (this%coo_rows(l) .GE. cur_row) EXIT
+                  l = l + 1
+               ENDDO
+               l_limit_left = l
+               IF (l .LE. l_limit_right) THEN
+                  IF (this%coo_rows(l) .EQ. cur_row) THEN
+                     ! We found a valid block. Figure out what to do with it.
+                     ! IF (DEBUG) print *, "Found row", this%coo_rows(l)
+                     IF (DEBUG) PRINT *, "(", cur_row, ",", cur_col, ")", l
+                     CALL dbcsr_get_stored_coordinates(matrix=this%dbcsr_mat, row=this%coo_rows(l), &
+                                                       column=this%coo_cols(l), processor=m)
+                     CALL blocks_from_rank_t(mytid)%sets(m)%insert(l)
+                  ENDIF
+               ENDIF
+            ENDDO
+         ENDDO
+      ENDDO
+      !$OMP END DO
+      !$OMP END PARALLEL
+
+      ! Merge partial results from threads
+      DO i = 0, this%numnodes - 1
+         DO j = 0, numthreads - 1
+            DO k = 1, result_blocks_from_rank_t(j)%sets(i)%elements
+               CALL this%result_blocks_from_rank(i)%insert(result_blocks_from_rank_t(j)%sets(i)%get(k))
+            ENDDO
+            CALL result_blocks_from_rank_t(j)%sets(i)%reset
+            DO k = 1, result_blocks_for_rank_t(j)%sets(i)%elements
+               CALL this%result_blocks_for_rank(i)%insert(result_blocks_for_rank_t(j)%sets(i)%get(k))
+            ENDDO
+            CALL result_blocks_for_rank_t(j)%sets(i)%reset
+            DO k = 1, blocks_from_rank_t(j)%sets(i)%elements
+               CALL blocks_from_rank(i)%insert(blocks_from_rank_t(j)%sets(i)%get(k))
+            ENDDO
+            CALL blocks_from_rank_t(j)%sets(i)%reset
+         ENDDO
+      ENDDO
+      DO i = 0, numthreads - 1
+         CALL nonzero_rows_t(i)%sets(1)%reset
+         DEALLOCATE (result_blocks_from_rank_t(i)%sets, result_blocks_for_rank_t(i)%sets, blocks_from_rank_t(i)%sets, &
+                     nonzero_rows_t(i)%sets)
+      ENDDO
+      DEALLOCATE (result_blocks_from_rank_t, result_blocks_for_rank_t, blocks_from_rank_t, nonzero_rows_t)
+
+      ! Make other ranks aware of our needs
+      ALLOCATE (num_blockids_send(0:this%numnodes - 1), num_blockids_recv(0:this%numnodes - 1))
+      DO i = 0, this%numnodes - 1
+         num_blockids_send(i) = blocks_from_rank(i)%elements
+      ENDDO
+      IF (DEBUG) PRINT *, this%myrank, "Number of block ids to send to each rank:", num_blockids_send
+#ifdef __parallel
+      CALL mp_alltoall(num_blockids_send, num_blockids_recv, 1, this%group)
+#else
+      num_blockids_recv(:) = num_blockids_send
+#endif
+      IF (DEBUG) PRINT *, this%myrank, "Number of block ids to receive from each rank:", num_blockids_recv
+      DO i = 0, this%numnodes - 1
+         CALL blocks_for_rank(i)%alloc(num_blockids_recv(i))
+      ENDDO
+      DEALLOCATE (num_blockids_send, num_blockids_recv)
+#ifdef __parallel
```

Comment:
It should not be necessary to have `#ifdef __parallel` in such high-level code. Our MPI wrappers will handle the serial case.

Reply:
I am indeed able to get rid of those around `mp_allgather` and `mp_alltoall`. However, `mp_isend` and `mp_irecv` raise an error unless `__parallel` is defined:
https://github.com/cp2k/cp2k/blob/master/src/mpiwrap/message_passing.f90#L3015

Reply:
Then you can use something like
`if (this%numnodes > 1)
   ... code with mp_isend/mp_irecv...
else
   sendbuffer = recvbuffer
end if`
That's what is done for instance in the MP2 code.